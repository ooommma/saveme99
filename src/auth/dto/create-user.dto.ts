import { IsEmail, IsString, MaxLength, MinLength } from 'class-validator';

export class CreateUserDto {
  @IsString()
  @IsEmail()
  email: string;

  @IsString()
  @MaxLength(15)
  @MinLength(4)
  password: string;

  profileImg?: string;

  @IsString()
  @MaxLength(10)
  @MinLength(2)
  name: string;
}
