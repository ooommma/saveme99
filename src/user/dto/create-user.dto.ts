import {
  IsEmail,
  IsNotEmpty,
  IsString,
  MaxLength,
  MinLength,
} from 'class-validator';

export class CreateUserDto {
  @IsString()
  @IsEmail()
  @IsNotEmpty()
  email: string;

  @IsString()
  @MaxLength(15)
  @MinLength(4)
  @IsNotEmpty()
  password: string;

  profileImg?: string;

  @IsString()
  @MaxLength(10)
  @MinLength(2)
  @IsNotEmpty()
  name: string;
}
