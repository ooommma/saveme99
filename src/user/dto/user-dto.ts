import { Users } from '../entities/users.entity';

export class UserDto {
  userId: number;
  email: string;
  profileImg: string;
  name: string;

  constructor(user: Users) {
    this.userId = user.userId;
    this.email = user.email;
    this.profileImg = user.profileImg;
    this.name = user.name;
  }
}
