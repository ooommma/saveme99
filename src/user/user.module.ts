import { Module, forwardRef } from '@nestjs/common';
import { UserService } from './user.service';
import { UserController } from './user.controller';
import { Users } from './entities/users.entity';
import { TypeOrmModule } from '@nestjs/typeorm';
import { AuthModule } from 'src/auth/auth.module';
import { AuthService } from 'src/auth/auth.service';
import { AwsModule } from 'src/aws/aws.module';
import { UtilsModule } from 'src/utils/utils.module';
import { JwtModule } from '@nestjs/jwt';

@Module({
  imports: [TypeOrmModule.forFeature([Users]), AwsModule, UtilsModule, forwardRef(() => AuthModule), JwtModule],
  controllers: [UserController],
  providers: [UserService, AuthService],
  exports: [UserService],
})
export class UserModule {}
