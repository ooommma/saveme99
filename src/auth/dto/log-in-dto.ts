import { PartialType, PickType } from '@nestjs/mapped-types';
import { CreateUserDto } from './create-user.dto';

export class UserLoginDto extends PickType(CreateUserDto, ['email', 'password']) {}
