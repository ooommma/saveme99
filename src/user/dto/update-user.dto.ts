import { PartialType, PickType } from '@nestjs/mapped-types';
import { CreateUserDto } from './create-user.dto';
import { ImageUploadDto } from 'src/aws/dto/image-upload.dto';
import { IsNotEmpty, IsString, MaxLength, MinLength } from 'class-validator';

export class UpdateUserDto extends ImageUploadDto {
  @IsString()
  @MaxLength(10)
  @MinLength(2)
  @IsNotEmpty()
  name: string;
}
