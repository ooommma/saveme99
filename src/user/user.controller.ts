import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  Put,
  UseGuards,
  UseInterceptors,
  HttpCode,
  UploadedFile,
} from '@nestjs/common';
import { UserService } from './user.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { AuthGuard } from '@nestjs/passport';
import { GetUser } from '../auth/decorator/get-user.decorator';
import { Users } from './entities/users.entity';
import { FileInterceptor } from '@nestjs/platform-express';
import {
  ApiBody,
  ApiConsumes,
  ApiOperation,
  ApiParam,
  ApiTags,
} from '@nestjs/swagger';
@ApiTags('유저 정보 수정/삭제')
@UseGuards(AuthGuard('jwt'))
@Controller('users')
export class UserController {
  constructor(private readonly userService: UserService) {}
  @UseInterceptors(FileInterceptor('file'))
  @ApiConsumes('multipart/form-data')
  @ApiBody({
    description: '수정할 데이터',
    schema: {
      type: 'object',
      properties: {
        name: { type: 'string' },
        file: {
          type: 'string',
          format: 'binary',
        },
      },
    },
  })
  @ApiOperation({ summary: '내 정보 수정' })
  @HttpCode(200)
  @Put('/:userId')
  async updateUser(
    @Body() updateUserDto: UpdateUserDto,
    @GetUser() user: Users,
    @Param('userId') userId: number,
    @UploadedFile() file: Express.Multer.File,
  ): Promise<Users> {
    return await this.userService.updateUser(updateUserDto, user, userId, file);
  }

  @ApiParam({
    name: 'userId',
    type: 'number',
    description: '사용자 ID',
  })
  @ApiOperation({ summary: '회원탈퇴' })
  @Delete('/:userId')
  async deleteUser(
    @Param('userId') userId: number,
    @GetUser() user: Users,
  ): Promise<string> {
    return await this.userService.deleteUser(userId, user);
  }
}
