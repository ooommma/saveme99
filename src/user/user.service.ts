import {
  ConflictException,
  Inject,
  Injectable,
  InternalServerErrorException,
  NotFoundException,
  UnauthorizedException,
  forwardRef,
} from '@nestjs/common';
import { UpdateUserDto } from './dto/update-user.dto';
import { Repository } from 'typeorm';
import { Users } from './entities/users.entity';
import { InjectRepository } from '@nestjs/typeorm';
import { AuthService } from 'src/auth/auth.service';

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(Users)
    private userRepository: Repository<Users>,
    @Inject(forwardRef(() => AuthService))
    private authService: AuthService,
  ) {}
  async updateUser(
    updateUserDto: UpdateUserDto,
    user: Users,
    userId: number,
    file: Express.Multer.File,
  ): Promise<Users> {
    const findUser: Users = await this.findUserById(userId);
    const { name, profileImg } = findUser;

    if (findUser.userId !== user.userId) {
      throw new UnauthorizedException('수정할 권한이 없습니다.');
    }
    if (name === updateUserDto.name) {
      throw new ConflictException('동일한 닉네임으로 변경할 수 없습니다.');
    }
    //Dto에 명시된 값 중 undefined인 값을 필터링
    const filteredDto = Object.entries(updateUserDto).reduce((acc, [key, value]) => {
      if (value !== undefined) {
        acc[key] = value;
      }
      return acc;
    }, {});
    //파일을 첨부하지 않았을 경우의 기존 이미지
    let imageUrl = profileImg;

    if (file) {
      const uploadedImageUrl = await this.authService.saveImage(file);
      imageUrl = uploadedImageUrl.imageUrl;
    }
    const updatedData = { ...findUser, ...filteredDto, profileImg: imageUrl };

    try {
      const updatedUser = await this.userRepository.save(updatedData);
      return updatedUser;
    } catch {
      throw new InternalServerErrorException('사용자 정보 수정에 실패했습니다.');
    }
  }

  async deleteUser(userId: number, user: Users): Promise<string> {
    const findUser = await this.findUserById(userId);
    if (findUser.userId !== user.userId) {
      throw new UnauthorizedException('회원탈퇴할 권한이 없습니다.');
    }

    const result = await this.userRepository.delete({ userId });
    if (result.affected === 0) {
      throw new NotFoundException(`${userId}번 유저가 존재하지 않아 삭제에 실패했습니다.`);
    }
    return `${userId}번 유저가 성공적으로 삭제되었습니다.`;
  }

  async findUserByEmail(email: string): Promise<Users> {
    console.log(email);
    const user = await this.userRepository.findOneBy({ email });
    if (!user) throw new NotFoundException(`${email}에 해당하는 유저를 찾을 수 없습니다.`);
    return user;
  }

  async findUserById(userId: number): Promise<Users> {
    const user = await this.userRepository.findOneBy({ userId });
    if (!user) throw new NotFoundException(`아이디가 ${userId}인 유저를 찾을 수 없습니다.`);
    return user;
  }

  async findUserByIdWithPassword(userId: number): Promise<Users> {
    const user = await this.userRepository.findOne({
      where: { userId },
      select: ['email', 'password', 'userId', 'createdAt', 'profileImg', 'updatedAt'],
    });
    if (!user) throw new NotFoundException(`아이디가 ${userId}인 유저를 찾을 수 없습니다.`);
    return user;
  }

  async findUserByEmailWithPassword(email: string): Promise<Users> {
    const user = await this.userRepository.findOne({
      where: { email },
      select: ['email', 'password', 'userId'],
    });
    if (!user) throw new NotFoundException(`${email}에 해당하는 유저를 찾을 수 없습니다.`);
    return user;
  }

  remove(id: number) {
    return `This action removes a #${id} user`;
  }
}
