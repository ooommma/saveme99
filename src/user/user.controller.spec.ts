import { Test, TestingModule } from '@nestjs/testing';
import { UserController } from './user.controller';
import { UserService } from './user.service';
import { UpdateUserDto } from './dto/update-user.dto';
import { Readable } from 'stream';
import { Users } from './entities/users.entity';
import { validate } from 'class-validator';

const mockUserService = {
  updateUser: jest.fn(),
  deleteUser: jest.fn(),
};

const updateUserDto: UpdateUserDto = {
  name: '김라임아님',
};

const mockFile: Express.Multer.File = {
  fieldname: 'file',
  originalname: 'testfile.txt',
  encoding: '7bit',
  mimetype: 'text/plain',
  size: 128,
  destination: './upload',
  filename: 'testfile.txt',
  path: './upload/testfile.txt',
  buffer: Buffer.from('Hello World'),
  stream: Readable.from(Buffer.from('Hello World')),
};

const mockUser: Users = {
  userId: 1,
  email: 'popcon940620@gmail.com',
  password: '42412412',
  name: '김라임',
  profileImg: null,
  createdAt: new Date(),
  updatedAt: new Date(),
  board: [],
  invitedBoards: [],
  comments: [],
};

describe('UserController', () => {
  let userController: UserController;
  let userService: UserService;

  beforeEach(async () => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();

    const module: TestingModule = await Test.createTestingModule({
      controllers: [UserController],
      providers: [
        {
          provide: UserService,
          useValue: mockUserService,
        },
      ],
    }).compile();

    userController = module.get<UserController>(UserController);
    userService = module.get<UserService>(UserService);
  });

  afterAll(async () => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should be defined', () => {
    expect(userController).toBeDefined();
    expect(userService).toBeDefined();
  });

  describe('updateUser Method', () => {
    it('success updateUser', async () => {
      const updateUserReturnVal = 'im updated user';

      mockUserService.updateUser.mockResolvedValue(updateUserReturnVal);

      const user = await userController.updateUser(
        updateUserDto,
        mockUser,
        1,
        mockFile,
      );

      expect(mockUserService.updateUser).toHaveBeenCalledWith(
        updateUserDto,
        mockUser,
        1,
        mockFile,
      );
      expect(mockUserService.updateUser).toHaveBeenCalledTimes(1);
      expect(user).toEqual(updateUserReturnVal);
    });
    it('updateUser should be failed by validate', async () => {
      const dto = new UpdateUserDto();
      dto.name = '김';
      const errors = await validate(dto);
      expect(errors.length).toBeGreaterThan(0);
      expect(errors).toMatchSnapshot();
    });
  });

  describe('deleteUser Method', () => {
    it('success deleteUser', async () => {
      const deleteMethodReturnVal = 'delete success';

      mockUserService.deleteUser.mockResolvedValue(deleteMethodReturnVal);

      const user = await userController.deleteUser(1, mockUser);

      expect(mockUserService.deleteUser).toHaveBeenCalledWith(1, mockUser);
      expect(mockUserService.deleteUser).toHaveBeenCalledTimes(1);
      expect(user).toEqual(deleteMethodReturnVal);
    });
  });
});
