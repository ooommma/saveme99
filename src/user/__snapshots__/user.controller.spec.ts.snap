// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController createUser Method signUp should be failed by validate 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "minLength": "name must be longer than or equal to 2 characters",
    },
    "property": "name",
    "target": UpdateUserDto {
      "name": "김",
    },
    "value": "김",
  },
]
`;

exports[`UserController updateUser Method updateUser should be failed by validate 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "minLength": "name must be longer than or equal to 2 characters",
    },
    "property": "name",
    "target": UpdateUserDto {
      "name": "김",
    },
    "value": "김",
  },
]
`;
